### PR TITLE
Rewrite in Go: Common reworks

### PR DESCRIPTION
**Pull request by dokshina**

### CODE REVIEW

File: src/build/build.go [hunk 1]
```diff
@@ -81,6 +68,10 @@ func checkCtx(projectCtx *project.ProjectCtx) error {
 	}
 
 	if projectCtx.BuildInDocker {
+		if projectCtx.Name == "" {
+			return fmt.Errorf("Name is missed")
```
Comment (olegrok) [inline src/build/build.go:72]: Is user in context what "Name" is missed?

Please add option name that is responsible for it

File: src/common/files.go [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+// IsExecOwner checks if specified file has owner execute permissions
+func IsExecOwner(path string) (bool, error) {
+	fileInfo, err := os.Stat(path)
+	if err != nil {
+		return false, err
+	}
+
+	perm := fileInfo.Mode().Perm()
+	return perm&0100 != 0, nil
```
Comment (olegrok) [inline src/common/files.go:21]: I suggest move "permission patterns" to constants

E.g. `perm&isOwner`... `perm&isExecReadWrite` and so on.

File: src/common/files.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+func FileLinesScanner(file *os.File) *bufio.Scanner {
+	scanner := bufio.NewScanner(file)
+	scanner.Split(bufio.ScanLines)
+	return scanner
+}
+
+// GetFileContent returns file content
+func GetFileContent(path string) (string, error) {
+	file, err := os.Open(path)
+	defer file.Close()
```
Comment (olegrok) [inline src/common/files.go:73]: Too early. Check for an error before and then your defer

File: src/pack/pack.go [hunk 4]
```diff
@@ -113,34 +117,50 @@ func Run(projectCtx *project.ProjectCtx) error {
 }
 
 func checkCtx(projectCtx *project.ProjectCtx) error {
-	if projectCtx.Path == "" {
-		return fmt.Errorf("Missed project path")
+	if projectCtx.Name == "" {
+		return fmt.Errorf("Path is missed")
```
Comment (olegrok) [inline src/pack/pack.go:121]: "Name is missed"?

File: src/pack/pack.go [hunk 5]
```diff
@@ -113,34 +117,50 @@ func Run(projectCtx *project.ProjectCtx) error {
...
+			projectCtx.SDKPath = projectCtx.TarantoolDir
 		}
 	}
+
+	return nil
 }
+
+const (
+	sdkPathError = `For packing in docker you should specify one of:
+	* --sdk-local: to use local SDK;;
```
Comment (olegrok) [inline src/pack/pack.go:163]: Redundant semicolon

File: src/pack/pack.go [hunk 6]
```diff
@@ -113,34 +117,50 @@ func Run(projectCtx *project.ProjectCtx) error {
...
+	* --sdk-path: path to SDK
+	  (can be passed in environment variable TARANTOOL_SDK_PATH)`
```
Comment (olegrok) [inline src/pack/pack.go:165]: Finish line with dot if you use semicolons for lines above.

File: test/python/test_pack.py [hunk 7]
```diff
@@ -445,7 +445,7 @@ def test_tempdir(cartridge_cmd, project_without_dependencies, tmpdir, pack_forma
     assert re.search(r'Temporary directory is set to {}'.format(cartridge_tempdir), output) is not None
 
 
-@pytest.mark.parametrize('pack_format', ['tgz', 'deb'])
```
Comment (olegrok) [inline test/python/test_pack.py:448]: Wy do you drop "deb"?

File: src/build/build.go (hunk 1 above)
Comment (dokshina) [reply src/build/build.go:72]: The context checks are internal checks. It's an internal error (like `check_state` in the old CLI)

File: test/python/test_pack.py (hunk 7 above)
Comment (dokshina) [reply test/python/test_pack.py:448]: Because this part is common for all types.

File: src/common/files.go (hunk 2 above)
Comment (dokshina) [reply src/common/files.go:21]: Fixed

File: cli/common/compress.go [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+		if fileInfo.Mode().IsRegular() {
+			if err := writeFileToWriter(filePath, tarWriter); err != nil {
+				return err
+			}
+		}
+
+		return nil
+	})
+
+	if err != nil {
```
Comment (olegrok) [inline cli/common/compress.go:46]: simply `return err`. If is redundant.

File: cli/common/compress.go [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+	destFile, err := os.Create(destFilePath)
+	if err != nil {
+		return fmt.Errorf("Failed to create result TGZ file %s: %s", destFilePath, err)
+	}
+
+	gzipWriter := gzip.NewWriter(destFile)
+	defer gzipWriter.Close()
+
+	err = WriteTarArchive(srcDirPath, gzipWriter)
+	if err != nil {
```
Comment (olegrok) [inline cli/common/compress.go:64]: if is redundant

File: cli/common/tarantool.go [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+	if tarantoolVersion == "" {
+		return "", fmt.Errorf("Failed to match Tarantool version")
+	}
+
+	return tarantoolVersion, nil
+}
+
+// GetNextMajorVersion computes next major version for a given one.
+// For example, for 1.10.3 it's 2
+func GetNextMajorVersion(version string) (string, error) {
```
Comment (olegrok) [inline cli/common/tarantool.go:70]: May be you can use "Segments" from go-version - https://github.com/hashicorp/go-version/blob/master/version.go#L330

File: cli/common/utils.go [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+		if value {
+			trueValuesCount++
+			if trueValuesCount > 1 {
+				return false
+			}
+		}
+	}
+
+	return trueValuesCount == 1
+
```
Comment (olegrok) [inline cli/common/utils.go:105]: redundant empty line

File: cli/pack/pack.go [hunk 12]
```diff
@@ -140,34 +145,50 @@ func Run(projectCtx *project.ProjectCtx) error {
 }
 
 func checkCtx(projectCtx *project.ProjectCtx) error {
-	if projectCtx.Path == "" {
-		return fmt.Errorf("Missed project path")
+	if projectCtx.Name == "" {
+		return fmt.Errorf("Path is missed")
```
Comment (olegrok) [inline cli/pack/pack.go:149]: Path is not Name

File: cli/pack/pack.go [hunk 13]
```diff
@@ -140,34 +145,50 @@ func Run(projectCtx *project.ProjectCtx) error {
...
 	}
 
-	if projectCtx.TarantoolDir == "" {
-		return fmt.Errorf("Missed Tarantool directory path")
+	if projectCtx.Path == "" {
+		return fmt.Errorf("Name is missed")
```
Comment (olegrok) [inline cli/pack/pack.go:153]: name is not path

File: test/python/test_pack.py (hunk 7 above)
Comment (olegrok) [reply test/python/test_pack.py:448]: Is it really consume a lot of time?

It's just for now and easily could be changed in future

File: cli/common/tarantool.go (hunk 10 above)
Comment (dokshina) [reply cli/common/tarantool.go:70]: Yes, thx.

Done.

**Pull request merged**